### PR TITLE
chore(enhancement): Fix pod autoscaler recovery and add total chaos duration as upper limit

### PR DESCRIPTION
**Pull request by uditgaurav**
Signed-off-by: Udit Gaurav <udit.gaurav@mayadata.io>
- In the pod autoscaler recovery, we should use the initial replica count rather than what is provided. 
- Add total chaos duration as the upper limit.

### CODE REVIEW

**Comment by uditgaurav** [comment]
- _When the chaos duration gets completed but the application replicas have still not scaled._ 

```bash
INFO[2020-11-03T21:06:59+05:30] Experiment Name: pod-autoscaler              
INFO[2020-11-03T21:06:59+05:30] [PreReq]: Getting the ENV for the  experiment 
INFO[2020-11-03T21:06:59+05:30] [PreReq]: Updating the chaos result of pod-autoscaler experiment (SOT) 
INFO[2020-11-03T21:07:02+05:30] The application informations are as follows   AppKind=deployment Label="run=nginx" Ramp Time=0 Namespace=default
INFO[2020-11-03T21:07:02+05:30] [Status]: Verify that the AUT (Application Under Test) is running (pre-chaos) 
INFO[2020-11-03T21:07:02+05:30] [Status]: Checking whether application containers are in running state 
INFO[2020-11-03T21:07:02+05:30] [Status]: The running status of container are as follows  container=nginx Pod=nginx-7bb7cd8db5-zzhkt Status=Running
INFO[2020-11-03T21:07:04+05:30] [Status]: Checking whether application pods are in running state 
INFO[2020-11-03T21:07:05+05:30] [Status]: The running status of Pods are as follows  Pod=nginx-7bb7cd8db5-zzhkt Status=Running
INFO[2020-11-03T21:07:08+05:30] Application Started Scaling                  
INFO[2020-11-03T21:07:08+05:30] Deployment's Available Replica Count is 1    
ERROR[2020-11-03T21:07:10+05:30] [Chaos]: Fail to scaleup the application replicas in the given chaos duration 
INFO[2020-11-03T21:07:13+05:30] [Info]: Application pod started rolling back 
INFO[2020-11-03T21:07:15+05:30] [RollBack]: Application Pod roll back to initial number of replicas 
ERROR[2020-11-03T21:07:15+05:30] Chaos injection failed, err: Unable to perform autoscaling, err: Status Check failed, err: Failed to scale the application 
```

- _When the application replicas are scaled but the chaos duration has completed yet._

```bash
INFO[2020-11-03T21:08:36+05:30] Experiment Name: pod-autoscaler              
INFO[2020-11-03T21:08:36+05:30] [PreReq]: Getting the ENV for the  experiment 
INFO[2020-11-03T21:08:36+05:30] [PreReq]: Updating the chaos result of pod-autoscaler experiment (SOT) 
INFO[2020-11-03T21:08:39+05:30] The application informations are as follows   Label="run=nginx" Ramp Time=0 Namespace=default AppKind=deployment
INFO[2020-11-03T21:08:39+05:30] [Status]: Verify that the AUT (Application Under Test) is running (pre-chaos) 
INFO[2020-11-03T21:08:39+05:30] [Status]: Checking whether application containers are in running state 
INFO[2020-11-03T21:08:39+05:30] [Status]: The running status of container are as follows  container=nginx Pod=nginx-7bb7cd8db5-zzhkt Status=Running
INFO[2020-11-03T21:08:41+05:30] [Status]: Checking whether application pods are in running state 
INFO[2020-11-03T21:08:41+05:30] [Status]: The running status of Pods are as follows  Pod=nginx-7bb7cd8db5-zzhkt Status=Running
INFO[2020-11-03T21:08:44+05:30] Application Started Scaling                  
INFO[2020-11-03T21:08:44+05:30] Deployment's Available Replica Count is 1    
INFO[2020-11-03T21:08:47+05:30] Deployment's Available Replica Count is 5    
INFO[2020-11-03T21:08:49+05:30] [Wait]: Waiting for completion of chaos duration 
INFO[2020-11-03T21:08:54+05:30] [Info]: Application pod started rolling back 
INFO[2020-11-03T21:08:57+05:30] [RollBack]: Application Pod roll back to initial number of replicas 
INFO[2020-11-03T21:08:57+05:30] [Confirmation]: The application pod autoscaler completed successfully 
INFO[2020-11-03T21:08:57+05:30] [Status]: Verify that the AUT (Application Under Test) is running (post-chaos) 
INFO[2020-11-03T21:08:57+05:30] [Status]: Checking whether application containers are in running state 
INFO[2020-11-03T21:09:09+05:30] [Status]: The running status of container are as follows  container=nginx Pod=nginx-7bb7cd8db5-zzhkt Status=Running
INFO[2020-11-03T21:09:11+05:30] [Status]: Checking whether application pods are in running state 
INFO[2020-11-03T21:09:12+05:30] [Status]: The running status of Pods are as follows  Pod=nginx-7bb7cd8db5-zzhkt Status=Running
INFO[2020-11-03T21:09:14+05:30] [The End]: Updating the chaos result of pod-autoscaler experiment (EOT) 
```

**Review by ksatchit — APPROVED** [review]

**Comment by uditgaurav** [comment]
```bash
INFO[2020-11-13T16:33:16+05:30] Experiment Name: pod-autoscaler              
INFO[2020-11-13T16:33:16+05:30] [PreReq]: Getting the ENV for the  experiment 
INFO[2020-11-13T16:33:16+05:30] [PreReq]: Updating the chaos result of pod-autoscaler experiment (SOT) 
INFO[2020-11-13T16:33:19+05:30] The application informations are as follows   AppKind=deployment Label="run=nginx" Ramp Time=0 Namespace=default
INFO[2020-11-13T16:33:19+05:30] [Status]: Verify that the AUT (Application Under Test) is running (pre-chaos) 
INFO[2020-11-13T16:33:19+05:30] [Status]: Checking whether application containers are in running state 
INFO[2020-11-13T16:33:19+05:30] [Status]: The running status of container are as follows  Pod=nginx-6db489d4b7-jndpj Status=Running container=nginx
INFO[2020-11-13T16:33:21+05:30] [Status]: Checking whether application pods are in running state 
INFO[2020-11-13T16:33:21+05:30] [Status]: The running status of Pods are as follows  Pod=nginx-6db489d4b7-jndpj Status=Running
INFO[2020-11-13T16:33:24+05:30] Application Started Scaling                  
INFO[2020-11-13T16:33:25+05:30] Deployment's Available Replica Count is 1    
INFO[2020-11-13T16:33:27+05:30] Deployment's Available Replica Count is 4    
INFO[2020-11-13T16:33:29+05:30] Deployment's Available Replica Count is 5    
INFO[2020-11-13T16:33:31+05:30] [Wait]: Waiting for completion of chaos duration 
INFO[2020-11-13T16:33:39+05:30] [Info]: Application pod started rolling back 
INFO[2020-11-13T16:33:41+05:30] [RollBack]: Application Pod roll back to initial number of replicas 
INFO[2020-11-13T16:33:41+05:30] [Confirmation]: The application pod autoscaler completed successfully 
INFO[2020-11-13T16:33:41+05:30] [Status]: Verify that the AUT (Application Under Test) is running (post-chaos) 
INFO[2020-11-13T16:33:41+05:30] [Status]: Checking whether application containers are in running state 
INFO[2020-11-13T16:33:51+05:30] [Status]: The running status of container are as follows  Status=Running container=nginx Pod=nginx-6db489d4b7-jndpj
INFO[2020-11-13T16:33:53+05:30] [Status]: Checking whether application pods are in running state 
INFO[2020-11-13T16:33:53+05:30] [Status]: The running status of Pods are as follows  Status=Running Pod=nginx-6db489d4b7-jndpj
INFO[2020-11-13T16:33:55+05:30] [The End]: Updating the chaos result of pod-autoscaler experiment (EOT) 
```

**Review by ksatchit — APPROVED** [review]

File: chaoslib/litmus/pod-autoscaler/lib/pod-autoscaler.go [hunk 1]
```diff
@@ -215,25 +218,30 @@ func DeploymentStatusCheck(experimentsDetails *experimentTypes.ExperimentDetails
...
 	}
 
-	// Keeping a wait time of 10s after all pod comes in running state
-	// This is optional and used just for viewing the pod status
-	time.Sleep(10 * time.Second)
+	//ChaosCurrentTimeStamp contains the current timestamp
+	ChaosCurrentTimeStamp := time.Now().Unix()
+	if int(ChaosCurrentTimeStamp-ChaosStartTimeStamp) <= experimentsDetails.ChaosDuration {
+		log.Info("[Wait]: Waiting for completion of chaos duration")
+		time.Sleep(time.Duration(int(ChaosCurrentTimeStamp-ChaosStartTimeStamp)) * time.Second)
```
Comment (ispeakc0de) [inline chaoslib/litmus/pod-autoscaler/lib/pod-autoscaler.go:230]: time.Sleep(time.Duration((experimentsDetails.ChaosDuration-int(ChaosCurrentTimeStamp-ChaosStartTimeStamp)) * time.Second)

File: chaoslib/litmus/pod-autoscaler/lib/pod-autoscaler.go [hunk 2]
```diff
@@ -254,14 +262,17 @@ func StatefulsetStatusCheck(experimentsDetails *experimentTypes.ExperimentDetail
...
```
Comment (ispeakc0de) [inline chaoslib/litmus/pod-autoscaler/lib/pod-autoscaler.go:274]: same as above

**Review by ispeakc0de — APPROVED** [review]

**Review by ksatchit — APPROVED** [review]

**Pull request merged**